File: packages/react-native-hms/ios/HMSInteractivityHelper.swift
//
//  HMSInteractivityHelper.swift
//  react-native-hms
//
//  Created by Jatin Nagar on 09/01/24.
//

import Foundation
import HMSSDK

class HMSInteractivityHelper {
    static func getPollBuilderFromDict(_ data: NSDictionary, sdkRoles: [HMSRole]?) -> HMSPollBuilder {
        let pollBuilder = HMSPollBuilder()

        if let title = data["title"] as? String {
            _ = pollBuilder.withTitle(title)
        }

        if let anonymous = data["anonymous"] as? Bool {
            _ = pollBuilder.withAnonymous(anonymous)
        }

        if let duration = data["duration"] as? Int {
            _ = pollBuilder.withDuration(duration)
        }

        if let category = data["type"] as? Int {
            if let pollCategory = HMSPollCategory.init(rawValue: category) {
                _ = pollBuilder.withCategory(pollCategory)
            } else {
                print("unable to create `HMSPollCategory` instance from given `type` value: \(category)")
            }
        }

        if let pollID = data["pollId"] as? String {
            _ = pollBuilder.withPollID(pollID)
        }

        if let rolesStrThatCanViewResponses = data["rolesThatCanViewResponses"] as? [String] {
            let roles = HMSHelper.getRolesFromRoleNames(rolesStrThatCanViewResponses, roles: sdkRoles)
            _ = pollBuilder.withRolesThatCanViewResponses(roles)
        }

        if let rolesStrThatCanVote = data["rolesThatCanVote"] as? [String] {
            let roles = HMSHelper.getRolesFromRoleNames(rolesStrThatCanVote, roles: sdkRoles)
            _ = pollBuilder.withRolesThatCanVote(roles)
        }

        if let userTrackingMode = data["mode"] as? Int {
            if let pollUserTrackingMode = HMSPollUserTrackingMode.init(rawValue: userTrackingMode) {
                _ = pollBuilder.withUserTrackingMode(pollUserTrackingMode)
            } else {
                print("unable to create `HMSPollUserTrackingMode` instance from given `userTrackingMode` value: \(userTrackingMode)")
            }
        }

        if let questions = data["questions"] as? [NSDictionary] {
            addQuestionsToPollBuilder(questions: questions, pollBuilder: pollBuilder)
        }

        return pollBuilder
    }

    static func addQuestionsToPollBuilder(questions: [NSDictionary], pollBuilder: HMSPollBuilder) {
        questions.forEach { question in
            let pollQuestionBuilder = self.getPollQuestionBuilderFromDict(question)
            _ = pollBuilder.addQuestion(with: pollQuestionBuilder)
        }
    }

    static func getPollQuestionBuilderFromDict(_ data: NSDictionary) -> HMSPollQuestionBuilder {
        let pollQuestionBuilder = HMSPollQuestionBuilder()

        // // Commented because as of now, not passing any field from JS to use `withAnswerHidden` method of `HMSPollQuestionBuilder`
        // if let answerHidden = data["answerHidden"] as? Bool {
        //     _ = pollQuestionBuilder.withAnswerHidden(answerHidden: answerHidden)
        // }

        if let canBeSkipped = data["skippable"] as? Bool {
            _ = pollQuestionBuilder.withCanBeSkipped(canBeSkipped)
        }

        if let once = data["once"] as? Bool {
            _ = pollQuestionBuilder.withCanChangeResponse(canChangeResponse: !once)
        }

        if let duration = data["duration"] as? Int {
            _ = pollQuestionBuilder.withDuration(duration)
        }

        if let index = data["index"] as? Int {
            _ = pollQuestionBuilder.withIndex(index)
        }

        if let maxLength = data["answerMaxLen"] as? Int {
            _ = pollQuestionBuilder.withMaxLength(maxLength: maxLength)
        }

        if let minLength = data["answerMinLen"] as? Int {
            _ = pollQuestionBuilder.withMinLength(minLength: minLength)
        }

        if let title = data["text"] as? String {
            _ = pollQuestionBuilder.withTitle(title)
        }

        if let type = data["type"] as? Int {
            if let questionType = HMSPollQuestionType.init(rawValue: type) {
                _ = pollQuestionBuilder.withType(questionType)
            } else {
                print("unable to create `HMSPollQuestionType` instance from given `type` value: \(type)")
            }
        }

        if let weight = data["weight"] as? Int {
            _ = pollQuestionBuilder.withWeight(weight: weight)
        }

        if let options = data["options"] as? [NSDictionary] {
            addOptionsToPollQuestionBuilder(options: options, pollQuestionBuilder: pollQuestionBuilder)
        }

        return pollQuestionBuilder
    }

    static func addOptionsToPollQuestionBuilder(options: [NSDictionary], pollQuestionBuilder: HMSPollQuestionBuilder) {
        options.forEach { option in
            guard let optionTitle = option["text"] as? String else {
                print("\(#function) option title field not available!")
                return
            }

            if let isCorrect = option["isCorrectAnswer"] as? Bool {
                _ = pollQuestionBuilder.addQuizOption(with: optionTitle, isCorrect: isCorrect)
            } else {
                _ = pollQuestionBuilder.addOption(with: optionTitle)
            }
        }
    }
}

File: packages/react-native-hms/android/src/main/java/com/reactnativehmssdk/Interactivity/HMSInteractivityHelper.kt
package com.reactnativehmssdk

import com.facebook.react.bridge.ReadableMap
import live.hms.video.polls.HMSPollBuilder
import live.hms.video.polls.HMSPollQuestionBuilder
import live.hms.video.polls.HMSPollResponseBuilder
import live.hms.video.polls.models.HmsPoll
import live.hms.video.polls.models.HmsPollCategory
import live.hms.video.polls.models.HmsPollUserTrackingMode
import live.hms.video.polls.models.question.HMSPollQuestion
import live.hms.video.polls.models.question.HMSPollQuestionType
import live.hms.video.sdk.models.role.HMSRole

object HMSInteractivityHelper {
  // region Create a poll
  fun getPollBuilder(
    data: ReadableMap,
    roles: List<HMSRole>,
  ): HMSPollBuilder {
    val pollBuilder = HMSPollBuilder.Builder()

    if (data.hasKey("isAnonymous")) {
      pollBuilder.withAnonymous(data.getBoolean("isAnonymous"))
    }

    if (data.hasKey("duration")) {
      pollBuilder.withDuration(data.getInt("duration").toLong())
    }

    if (data.hasKey("mode")) {
      pollBuilder.withUserTrackingMode(getUserTrackingMode(data))
    }

    if (data.hasKey("type")) {
      pollBuilder.withCategory(getPollCategory(data))
    } else {
      pollBuilder.withCategory(HmsPollCategory.POLL)
    }

    data.getString("pollId")?.let {
      pollBuilder.withPollId(it)
    }
    data.getString("title")?.let {
      pollBuilder.withTitle(it)
    }

    if (data.hasKey("rolesThatCanViewResponses")) {
      pollBuilder.withRolesThatCanViewResponses(getRolesThatCanViewResponses(data, roles))
    }

    if (data.hasKey("rolesThatCanVote")) {
      pollBuilder.withRolesThatCanVote(getRolesThatCanVote(data, roles))
    }

    if (data.hasKey("questions")) {
      val questions = data.getArray("questions")?.toArrayList() as? ArrayList<HashMap<String, Any>>

      if (questions != null) {
        addQuestions(questions, pollBuilder)
      }
    }

    return pollBuilder.build()
  }

  // endregion

  // region Poll Builder Helpers
  private fun getQuestionType(type: Int): HMSPollQuestionType {
    return when (type) {
      0 -> HMSPollQuestionType.singleChoice
      1 -> HMSPollQuestionType.multiChoice
      2 -> HMSPollQuestionType.shortAnswer
      3 -> HMSPollQuestionType.longAnswer
      else -> HMSPollQuestionType.singleChoice
    }
  }

  private fun getUserTrackingMode(data: ReadableMap): HmsPollUserTrackingMode {
    return when (data.getInt("mode")) {
      0 -> HmsPollUserTrackingMode.PEER_ID
      1 -> HmsPollUserTrackingMode.USER_ID
      2 -> HmsPollUserTrackingMode.USERNAME
      else -> HmsPollUserTrackingMode.PEER_ID
    }
  }

  fun getPollCategory(data: ReadableMap): HmsPollCategory {
    return when (data.getInt("type")) {
      0 -> HmsPollCategory.POLL
      1 -> HmsPollCategory.QUIZ
      else -> HmsPollCategory.POLL
    }
  }

  private fun getRolesThatCanViewResponses(
    data: ReadableMap,
    roles: List<HMSRole>,
  ): List<HMSRole> {
    val rolesThatCanViewResponsesString =
      data.getArray("rolesThatCanViewResponses")?.toArrayList() as? ArrayList<String>
    val encodedTargetedRoles =
      return HMSHelper.getRolesFromRoleNames(rolesThatCanViewResponsesString, roles)
  }

  private fun getRolesThatCanVote(
    data: ReadableMap,
    roles: List<HMSRole>,
  ): List<HMSRole> {
    val rolesThatCanVoteString =
      data.getArray("rolesThatCanVote")?.toArrayList() as? ArrayList<String>
    return HMSHelper.getRolesFromRoleNames(rolesThatCanVoteString, roles)
  }

  private fun addQuestions(
    questions: ArrayList<HashMap<String, Any>>,
    pollBuilder: HMSPollBuilder.Builder,
  ) {
    for (question in questions) {
      val type = question["type"] as? Double
      type?.let {
        addPollBuilderQuestion(getQuestionType(it.toInt()), question, pollBuilder)
      }
    }
  }

  private fun addSingleChoiceQuestion(
    item: HashMap<String, Any>,
    pollBuilder: HMSPollBuilder.Builder,
  ) {
    if (item["title"] != null && item["options"] != null) {
      val title = item["title"] as? String
      val options = item["options"] as? ArrayList<String>

      if (title != null && options != null) {
        val questionBuilder =
          HMSPollQuestionBuilder.Builder(HMSPollQuestionType.singleChoice)
            .withTitle(title)

        for (i in 0 until options.size) {
          questionBuilder.addOption(options[i])
        }
        pollBuilder.addQuestion(questionBuilder.build())
      }
    }
  }

  private fun addMultipleChoiceQuestion(
    item: HashMap<String, Any>,
    pollBuilder: HMSPollBuilder.Builder,
  ) {
    if (item["title"] != null && item["options"] != null) {
      val title = item["title"] as? String
      val options = item["options"] as? ArrayList<String>

      if (title != null && options != null) {
        val questionBuilder =
          HMSPollQuestionBuilder.Builder(HMSPollQuestionType.multiChoice)
            .withTitle(title)

        for (i in 0 until options.size) {
          questionBuilder.addOption(options[i])
        }
        pollBuilder.addQuestion(questionBuilder.build())
      }
    }
  }

  private fun addShortAnswerQuestion(
    item: HashMap<String, Any>,
    pollBuilder: HMSPollBuilder.Builder,
  ) {
    if (item["title"] != null) {
      val title = item["title"] as? String

      if (title != null) {
        val questionBuilder =
          HMSPollQuestionBuilder.Builder(HMSPollQuestionType.shortAnswer)
            .withTitle(title)
        pollBuilder.addQuestion(questionBuilder.build())
      }
    }
  }

  private fun addLongAnswerQuestion(
    item: HashMap<String, Any>,
    pollBuilder: HMSPollBuilder.Builder,
  ) {
    if (item["title"] != null) {
      val title = item["title"] as? String

      if (title != null) {
        val questionBuilder =
          HMSPollQuestionBuilder.Builder(HMSPollQuestionType.longAnswer)
            .withTitle(title)
        pollBuilder.addQuestion(questionBuilder.build())
      }
    }
  }

  private fun addPollBuilderQuestion(
    type: HMSPollQuestionType,
    item: HashMap<String, Any>,
    pollBuilder: HMSPollBuilder.Builder,
  ) {
    val questionBuilder = HMSPollQuestionBuilder.Builder(type)

//    val answerHidden = item["answerHidden"] as? Boolean
//    if (answerHidden != null) {
//      questionBuilder.withAnswerHidden(answerHidden)
//    }

    val canBeSkipped = item["skippable"] as? Boolean
    if (canBeSkipped != null) {
      questionBuilder.withCanBeSkipped(canBeSkipped)
    }

    val canChangeResponse = item["once"] as? Boolean
    if (canChangeResponse != null) {
      questionBuilder.withCanChangeResponse(!canChangeResponse)
    }

    val duration = item["duration"] as? Int
    if (duration != null) {
      questionBuilder.withDuration(duration.toLong())
    }

    val maxLength = item["answerMaxLen"] as? Int
    if (maxLength != null) {
      questionBuilder.withMaxLength(maxLength.toLong())
    }

    val minLength = item["answerMinLen"] as? Int
    if (minLength != null) {
      questionBuilder.withMinLength(minLength.toLong())
    }

    val title = item["text"] as? String
    if (title != null) {
      questionBuilder.withTitle(title)
    }

    val weight = item["weight"] as? Int
    if (weight != null) {
      questionBuilder.withWeight(weight)
    }

    val options = item["options"] as? ArrayList<HashMap<String, Any>>
    options?.let {
      addOptions(it, questionBuilder)
    }

    questionBuilder.build().let {
      pollBuilder.addQuestion(it)
    }
  }

  private fun addOptions(
    options: ArrayList<HashMap<String, Any>>,
    questionBuilder: HMSPollQuestionBuilder.Builder,
  ) {
    for (i in 0 until options.size) {
      val option = options[i]

      if (option["isCorrect"] != null) {
        val isCorrect = option["isCorrectAnswer"] as? Boolean
        if (isCorrect != null) {
          questionBuilder.addQuizOption(
            option["text"] as? String ?: "",
            isCorrect,
          )
        }
      } else {
        questionBuilder.addOption(option["text"] as? String ?: "")
      }
    }
  }

  // endregion

  // region Poll Response

  fun getPollResponseBuilder(
    response: ReadableMap,
    poll: HmsPoll,
    pollQuestion: HMSPollQuestion,
  ): HMSPollResponseBuilder {
    val pollResponseBuilder = HMSPollResponseBuilder(hmsPoll = poll, userId = null)

    when (pollQuestion.type) {
      HMSPollQuestionType.longAnswer, HMSPollQuestionType.shortAnswer -> {
        val responseText = response.getString("text")
        if (responseText != null) {
          if (response.hasKey("duration")) {
            val duration = response.getInt("duration")
            pollResponseBuilder.addResponse(pollQuestion, responseText, duration.toLong())
          } else {
            pollResponseBuilder.addResponse(pollQuestion, responseText)
          }
        }
      }
      HMSPollQuestionType.multiChoice, HMSPollQuestionType.singleChoice -> {
        val options = response.getArray("options")?.toArrayList() as? ArrayList<Int>
        val pollQuestionOptions = pollQuestion.options

        if (options != null && pollQuestionOptions != null) {
          val questionOptions =
            options.mapNotNull { optionIndex ->
              pollQuestionOptions.firstOrNull { pollQuestionOption ->
                pollQuestionOption.index == optionIndex
              }
            }

          if (response.hasKey("duration")) {
            val duration = response.getInt("duration")
            pollResponseBuilder.addResponse(pollQuestion, questionOptions, duration.toLong())
          } else {
            pollResponseBuilder.addResponse(pollQuestion, questionOptions)
          }
        }
      }
      else -> {
        // @unknown default case: Invalid pollQuestionType
      }
    }
    return pollResponseBuilder
  }

  // endregion
}
